### PR TITLE
coord: Fix read-then-write serializability issue

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, read-then-write statements would release the global write
lock, before queuing their writes for a group commit. This allowed two
read-then-write statements to interleave their reads and writes, which
can lead to a serializable consistency violation.

An example of one such violation is the following:
1. Txn 1 acquires lock
2. Txn 1 reads from A
3. Txn 1 releases lock
4. Txn 1 submites write to A
5. Txn 2 acquires lock
6. Txn 2 reads from A
7. Txn 2 releases lock
8. Txn 2 submits write to A
9. Group commit executes
There is no serial order that would be equivalent to this order.

This commit updates the behavior of read-then-write statements to hold
on to the global write lock until group commit executes. This means
that no two read-then-write statements can be part of the same group
commit.

However, we can include one read-then-write statements and many write
transactions in the same group commit without violating
serializability. Specifically, the serial order would be to order all
of the write transactions after the read-then-write statement.

### Motivation
This PR fixes a previously unreported bug.

### Testing

- [X] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Review by mjibson — COMMENTED** [review]
I *think* this looks correct (which I thought before too so...). I have a light preference for a test that would have caught the previous error. Feels like a thing that could be triggered by hammering with an append-only workload and a verifyer, like a janky hand written jepsen. Or maybe we should just do jepsen? Another test idea to which I'm not opposed: write a test that does this and looks for problems, but we ignore that test in CI for now? We previously had a linearizability test like that that we could uncomment when doing development. That may be good enough for now until we maybe have like a nightly jepsen thing going.

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -3546,11 +3575,11 @@ impl<S: Append + 'static> Coordinator<S> {
 
                     // `rows` can be empty if, say, a DELETE's WHERE clause had 0 results.
                     writes.retain(|WriteOp { rows, .. }| !rows.is_empty());
-                    return Ok(Some(writes));
+                    return Ok((Some(writes), write_lock_guard));
```
Comment (mjibson) [inline src/coord/src/coord.rs:3578]: Lock passing. Cool.

File: src/coord/src/coord.rs [hunk 2]
```diff
@@ -1231,11 +1261,9 @@ impl<S: Append + 'static> Coordinator<S> {
 
     /// Submit a write to be executed during the next group commit.
     fn submit_write(&mut self, pending_write_txn: PendingWriteTxn) {
-        if self.pending_writes.is_empty() {
-            self.internal_cmd_tx
-                .send(Message::GroupCommit)
-                .expect("sending to internal_cmd_tx cannot fail");
-        }
+        self.internal_cmd_tx
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/coord/src/coord.rs:1264]: This is part of the complexity that I wasn't a huge fan of. When a read-then-write transaction submits its writes, then we need to re-trigger a group commit even if there's already existing pending write transactions. The reason is that those write transactions are waiting on this read-then-write transaction to transfer ownership of the lock to `pending_writes`

**Comment by jkosh44** [comment]
> I _think_ this looks correct (which I thought before too so...). I have a light preference for a test that would have caught the previous error. Feels like a thing that could be triggered by hammering with an append-only workload and a verifyer, like a janky hand written jepsen. Or maybe we should just do jepsen? Another test idea to which I'm not opposed: write a test that does this and looks for problems, but we ignore that test in CI for now? We previously had a linearizability test like that that we could uncomment when doing development. That may be good enough for now until we maybe have like a nightly jepsen thing going.

I still have that linearizability test floating around locally, though I don't think that would have caught this. If I'm thinking of the right test, then that tests if we choose some earlier read timestamp when compared to a previous read. Where as this violation happens because we interleave reads from two transactions.

I can add that janky jepsen like test, I think it'd be pretty simple. Just spin up X number of threads that all execute `INSERT INTO foo SELECT * FROM foo`, then count the number of rows at the end. It wouldn't deterministically fail, but I think it would fail sometimes.

I do think eventually we want to try and add maelstrom testing to the Coord layer.

File: src/environmentd/tests/sql.rs [hunk 3]
```diff
@@ -1218,3 +1218,55 @@ fn test_tail_outlive_cluster() {
         0
     );
 }
+
+#[test]
```
Comment (jkosh44) [inline src/environmentd/tests/sql.rs:1222]: @mjibson This fails on main.

Comment (mjibson) [reply src/environmentd/tests/sql.rs:1222]: Lol it fails with only 3 loops 3 threads? Ha. Good test.

**Review by mjibson — APPROVED** [review]

**Pull request merged**